Feature/Update documentation for optional update IDs (#814)
PowerGridModel/power-grid-model: docs/user_manual/components.md

| name | data type | unit | description                                                                                                                                                       | required |                                     update                                     |
| ---- | --------- | ---- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- | :------: | :----------------------------------------------------------------------------: |
| `id` | `int32_t` | -    | ID of a component. The ID should be unique across all components within the same scenario, e.g., you cannot have a node with `id=5` and another line with `id=5`. | &#10004; | &#10060; (see below) |

If a component update is uniform and is updating all the elements with the same component type, IDs can be omitted or set to `nan`s. In any other case, the IDs need to be present in the update dataset, but cannot be changed.

> Uniform component updates are ones that update the same number of component elements across scenarios. By definition, a dense update is always uniform, hence the IDs can always be optional; whereas for a sparse update, the IDs can only be optional if the index pointer is given as an arithmetic sequence of all elements (i.e. a sparse representation of a dense buffer). An example of the usage of optional IDs is given in [Power Flow Example](./Power%20Flow%20Example.ipynb)

Reviewer: there is a bug when dealing with mixed optional ID, e.g.:

```py
update_sym_load_no_id = initialize_array(DatasetType.update, ComponentType.sym_load, (3, 2))
update_sym_load_no_id["p_specified"] = [30e6, 15e6]  # broadcast

update_line = initialize_array(DatasetType.update, ComponentType.line, (3, 1))
update_line["id"] = [[3], [5], [8]]  # broadcast
update_line["from_status"] = [0]

update_data_no_id = {ComponentType.sym_load: update_sym_load_no_id, ComponentType.line: update_line}
```

this will pass `assert_valid_batch_data` but will still throw `PowerGridBatchErrors`:
```
[IDNotFound('The id cannot be found: -2147483648\n'), IDNotFound('The id cannot be found: -2147483648\n'), IDNotFound('The id cannot be found: -2147483648\n')]
```

Reviewer: resolved in #819

> Uniform component updates are ones that update the same number of component elements across scenarios. By definition, a dense update is always uniform, hence the IDs can always be optional; whereas for a sparse update, the IDs can only be optional if the index pointer is given as an arithmetic sequence of all elements (i.e. a sparse representation of a dense buffer). An example of the usage of optional IDs is given in [Power Flow Example](./Power%20Flow%20Example.ipynb)

Reviewer: ```suggestion
Uniform component updates are ones that update the same number of component of the same type across scenarios. By definition, a dense update is always uniform, hence the IDs can always be optional; whereas for a sparse update, the IDs can only be optional if the index pointer is given as an arithmetic sequence of all elements (i.e. a sparse representation of a dense buffer). An example of the usage of optional IDs is given in [Power Flow Example](./Power%20Flow%20Example.ipynb)
```

| name | data type | unit | description                                                                                                                                                       | required |                                     update                                     |
| ---- | --------- | ---- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- | :------: | :----------------------------------------------------------------------------: |

> | `id` | `int32_t` | -    | ID of a component. The ID should be unique across all components within the same scenario, e.g., you cannot have a node with `id=5` and another line with `id=5`. | &#10004; | &#10060; (If a component update is uniform, the ID can be omitted or set to `nan`. Otherwise, the id needs to be specified in the update query, but cannot be changed) |

Reviewer: i don't know whether a user understands what `uniform` means if you put it in the components like this.

you can also just say something like `(see below)` and then have a note block elaborate on the topic

Author: Good idea, I will add a block further explaining below.

| name | data type | unit | description                                                                                                                                                       | required |                                     update                                     |
| ---- | --------- | ---- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- | :------: | :----------------------------------------------------------------------------: |

> | `id` | `int32_t` | -    | ID of a component. The ID should be unique across all components within the same scenario, e.g., you cannot have a node with `id=5` and another line with `id=5`. | &#10004; | &#10060; (If a component update is uniform (see below), the ID can be omitted or set to `nan`. Otherwise, the id needs to be specified in the update query, but cannot be changed) |

Reviewer 2: ```suggestion
| `id` | `int32_t` | -    | ID of a component. The ID should be unique across all components within the same scenario, e.g., you cannot have a node with `id=5` and another line with `id=5`. | &#10004; | &#10060; (if a component update is uniform (see below) and is updating all the elements with the same component type, IDs can be omitted or set to `nan`. In any other case, IDs need to be present in the update dataset for reference) |
```

Reviewer: i feel like this is too long. it's better to divert as soon to the comment below as possible and to keep the table as brief as possible

Reviewer 2: ```suggestion
| `id` | `int32_t` | -    | ID of a component. The ID should be unique across all components within the same scenario, e.g., you cannot have a node with `id=5` and another line with `id=5`. | &#10004; | &#10060; See below for optional IDs. |
```

> | `id` | `int32_t` | -    | ID of a component. The ID should be unique across all components within the same scenario, e.g., you cannot have a node with `id=5` and another line with `id=5`. | &#10004; | &#10060; (If a component update is uniform (see below), the ID can be omitted or set to `nan`. Otherwise, the id needs to be specified in the update query, but cannot be changed) |

Reviewer: the unicode for update &#10060; needs to be adjusted

Author: Deemed correct after discussion.

> A uniform component update is defined as one in which every scenario modifies the same number of elements within that component. Therefore, a dense update is always uniform, hence the ID can always be optional; whereas a sparse update is only uniform if the index pointer is given as a linear sequence of elements (sparse representation of a dense buffer). An example of the usage of optional IDs is given in [Power Flow Example](./Power%20Flow%20Example.ipynb)

Reviewer 2: ```suggestion
Uniform component updates ones that update the same number of component elements across scenarios. By definition, a dense update is always uniform, hence the ID can always be optional; ...
```

``` a sparse update is only uniform if the index pointer is given as a linear sequence of elements```

I don't think this is necessarily true. Sequence `{1, 2, 3}` is linear, just like `{2, 4, 6}` is linear, and `{6, 4, 2}`.

Reviewer 2: Maybe consider:
``` suggestion
... whereas a sparse update is only uniform if the index pointer is given as an arithmetic sequence of all elements (sparse representation of a dense buffer). An example of the usage of optional IDs is given in [Power Flow Example](./Power%20Flow%20Example.ipynb)
```

> A component update is uniform when the number of elements is consistent across all scenarios and matches the number of elements in the input data for that component. Therefore, the ID is not optional for updates on sparse data, and when the ID of a component is ommited, the updated field must be explicit across all scenarios (even if it is not always updated). An example of the usage of optional IDs is given in [Power Flow Example](./Power%20Flow%20Example.ipynb)

Reviewer 2: > is uniform when the number of elements is consistent across all scenarios and matches the number of elements in the input data

Not sure if this **was** the case. Updates can all be updating to the same components but not all the components. 

> Therefore, the ID is not optional for updates on sparse data, and when the ID of a component is ommited

Being `uniform` and being `sparse` are two orthogonal concepts I believe. @mgovers could you confirm this?

> the updated field must be explicit across all scenarios (even if it is not always updated).

I do not quite understand this line

Reviewer 2: Oh, it's not yet in review. My apologies.

Reviewer: > Being uniform and being sparse are two orthogonal concepts I believe. @mgovers could you confirm this?

see my post in teams in which i elaborate the existing conventions used in the docs 😬 if we can make clear boundaries on terminology, we may resolve this more straightforwardly

Reviewer: short version: there's a difference between the API reference and the implementation in main model and we need to agree on one set of conventions

Author: > > is uniform when the number of elements is consistent across all scenarios and matches the number of elements in the input data
> 
> Not sure if this **was** the case. Updates can all be updating to the same components but not all the components.

Maybe my explanation wasn't clear, but this intended to reflect the fact that even if you are updating one line (out of 3 for instance), you still need to provide the whole column of the updated attribute in the scenario where it is updated. My intention was to "demonstrate" this in the example notebook.

> > Therefore, the ID is not optional for updates on sparse data, and when the ID of a component is ommited
> 
> Being `uniform` and being `sparse` are two orthogonal concepts I believe. @mgovers could you confirm this?

I see this now, I was following the documentation rather than the code logic. I will update.

> > the updated field must be explicit across all scenarios (even if it is not always updated).
> 
> I do not quite understand this line

I had a hard time trying to write this down concisely and clearly, but it looks like I failed. The line tries to explain what I replied on the first line above.

> short version: there's a difference between the API reference and the implementation in main model and we need to agree on one set of conventions

We can discuss tomorrow, but I believe we should go with the code here.

Please do give me suggestions if you feel I could have written something better, specially regarding the uniform part.

Author: I have adjusted in [5bc7cf5](https://github.com/PowerGridModel/power-grid-model/pull/814/commits/5bc7cf5b7b6a24ef5b5540d85703f61db432e830) according to the team discussions about the definitions.
